<a href="https://colab.research.google.com/github/aydawudu/ML-Zoomcamp/blob/main/Homework1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#Question 1

What's the version of Numpy you installed?


In [ ]:
np.__version__

'1.21.6'

Getting the data

In [ ]:
df=pd.read_csv("/content/data.csv")

# Question 2
How many records are in the dataset?

Here you need to specify the number of rows.

In [ ]:
# Answer: There are 11914 rows
df.shape[0]

11914

#Question 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

In [ ]:
# Answer: Chevrolet, Ford, Volkswagen are the most frequent car manufacturers
df.Make.value_counts()


Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
Alfa Romeo          5
McLaren             5
Spyker    

#Question 4
What's the number of unique Audi car models in the dataset?

Answer=328


#Question 5
How many columns in the dataset have missing values?

In [ ]:
#Answer: There are 5 columns with missing value
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

#Question 6
Find the median value of "Engine Cylinders" column in the dataset.
Next, calculate the most frequent value of the same "Engine Cylinders".
Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
Now, calculate the median value of "Engine Cylinders" once again.
Has it changed?

In [ ]:
df['Engine Cylinders'].median()

6.0

In [ ]:
df['Engine Cylinders'].value_counts()

4.0     4752
6.0     4489
8.0     2031
12.0     230
5.0      225
10.0      68
0.0       56
3.0       30
16.0       3
Name: Engine Cylinders, dtype: int64

In [ ]:
df=df.fillna(4)
df['Engine Cylinders'].median()

6.0

In [ ]:
#Answer, the median has not changed

#Question 7
1. Select all the "Lotus" cars from the dataset.
2. Select only columns "Engine HP", "Engine Cylinders".
3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the value of the first element of w?

In [ ]:
df_lotus=df[(df["Make"]=='Lotus')] #Select all the "Lotus" cars from the dataset.
df_new=df_lotus[["Engine HP", "Engine Cylinders"]] #Select only columns "Engine HP", "Engine Cylinders".
X=df_new.drop_duplicates() #Now drop all duplicated rows using drop_duplicates method AND Get the underlying NumPy array. Let's call it X..
X=X.to_numpy()
XTX=X*X #Compute matrix-matrix multiplication between the transpose of X and X. 
XTX=XTX.T #Invert XTX.
y=np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]) #Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]
w=XTX * y #Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w[0]


array([3.929310e+07, 3.801920e+07, 3.531675e+07, 1.041250e+08,
       2.080000e+08, 7.617600e+07, 1.190250e+08, 8.586370e+07,
       4.608000e+07])